packages/block-editor/src/components/inserter/block-patterns.js

```diff
@@ -52,49 +81,130 @@ function BlockPatternPlaceholder() {
 	);
 }
 
-function BlockPatterns( { patterns, onInsert, filterValue } ) {
+function BlockPatternList( { patterns, shownPatterns, onClickPattern } ) {
+	return patterns.map( ( pattern ) => {
+		const isShown = shownPatterns.includes( pattern );
+		return isShown ? (
+			<BlockPattern
+				key={ pattern.name }
+				pattern={ pattern }
+				onClick={ onClickPattern }
+			/>
+		) : (
+			<BlockPatternPlaceholder key={ pattern.name } />
+		);
+	} );
+}
+
+function BlockPatternsSearchResults( { filterValue, onInsert } ) {
+	const [ patterns, , onClick ] = usePatternsState( onInsert );
+	const currentShownPatterns = useAsyncList( patterns );
+
 	const filteredPatterns = useMemo(
 		() => searchItems( patterns, filterValue ),
 		[ filterValue, patterns ]
 	);
-	const currentShownPatterns = useAsyncList( filteredPatterns );
-	const { createSuccessNotice } = useDispatch( 'core/notices' );
-	const onClickPattern = useCallback( ( pattern, blocks ) => {
-		onInsert( map( blocks, ( block ) => cloneBlock( block ) ) );
-		createSuccessNotice(
-			sprintf(
-				/* translators: %s: block pattern title. */
-				__( 'Pattern "%s" inserted.' ),
-				pattern.title
-			),
-			{
-				type: 'snackbar',
-			}
+
+	if ( filterValue ) {
+		return !! filteredPatterns.length ? (
+			<InserterPanel title={ __( 'Search Results' ) }>
+				<BlockPatternList
+					shownPatterns={ currentShownPatterns }
+					patterns={ filteredPatterns }
+					onClickPattern={ onClick }
+				/>
+			</InserterPanel>
+		) : (
+			<InserterNoResults />
 		);
-	}, [] );
+	}
+}
 
-	return !! filteredPatterns.length ? (
-		<InserterPanel
-			title={
-				filterValue
-					? __( 'Search Results' )
-					: _x( 'All', 'patterns categories' )
+function BlockPatternsPerCategories( { onInsert } ) {
+	const [ patterns, categories, onClick ] = usePatternsState( onInsert );
+
+	const getPatternIndex = useCallback(
+		( pattern ) => {
+			if ( ! pattern.categories || ! pattern.categories.length ) {
+				return Infinity;
 			}
-		>
-			{ filteredPatterns.map( ( pattern, index ) =>
-				currentShownPatterns[ index ] === pattern ? (
-					<BlockPattern
-						key={ pattern.name }
-						pattern={ pattern }
-						onClick={ onClickPattern }
-					/>
-				) : (
-					<BlockPatternPlaceholder key={ pattern.name } />
+			const indexedCategories = Object.fromEntries(
```

Comment:
`Object.fromEntries` is newer than the `wp-polyfill` polyfill provides. It's not polyfilled for Internet Explorer.

![image](https://user-images.githubusercontent.com/1779930/82094963-13ecc400-96cc-11ea-919d-18d77f769ebb.png)

```
TypeError: Object doesn't support property or method 'fromEntries'
   at Anonymous function (http://192.168.1.162:8888/wp-content/plugins/gutenberg/build/block-editor/index.js?ver=cfa28e6fe47f5d29787baf4893617371:20456:5)
   at Anonymous function (http://192.168.1.162:8888/wp-content/plugins/gutenberg/build/block-editor/index.js?ver=cfa28e6fe47f5d29787baf4893617371:20467:7)
   at Anonymous function (http://192.168.1.162:8888/wp-content/plugins/gutenberg/build/block-editor/index.js?ver=cfa28e6fe47f5d29787baf4893617371:20466:5)
   at mountMemo (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:15807:3)
   at HooksDispatcherOnMountInDEV.useMemo (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:16029:9)
   at useMemo (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react.js?ver=16.9.0:1680:3)
   at BlockPatternsPerCategories (http://192.168.1.162:8888/wp-content/plugins/gutenberg/build/block-editor/index.js?ver=cfa28e6fe47f5d29787baf4893617371:20465:3)
   at renderWithHooks (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:15246:3)
   at mountIndeterminateComponent (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:17480:5)
   at beginWork$1 (http://192.168.1.162:8888/wp-includes/js/dist/vendor/react-dom.js?ver=16.9.0:18624:9)
```

Reply:
Oh, TIL. `fromEntries` was my suggestion, and I didn't stop to think it was a recent addition.

Reply:
@youknowriad, we can replace this directly with `lodash.fromPairs`.

Reply:
yes, that's what I did in #22410

Reply:
Related: #22417